*****

# Segmenting and Clustering Neighborhoods in Toronto

Project: Ari Bimo Prakoso

# Table of Content

### Part 1: Getting Dataframe about Postal Code, Borough, and Neighborhoods
* Import libraries.
* Save the link of postal codes.
* Extract information from the website with BeautifulSoup.
* Preprocess data into dataframe.
* Process cells with defined borough.
* Combine multiple neighborhoods.
* Assign norough to 'Not Assigned' beighborhood.


### Part 2: Getting Dataframe with Latitude and Longitude Data
* Get the coordinates data.
* Get the latitude and longitude data, put it into dataframe by matching postal code.
* Dataframe ready for clustering.

### Part 3: Neighborhood Clustering

* Render Toronto map and each postal code.
* Cluster based on venues of interest.
* Explore first neighborhood.
* Explore all neighborhoods.
* Get data of venues in Toronto.
* Analyze neighborhoods.
* Use machine learning to cluster the neighborhood.
* Plot the cluster on Folium map.
* Examine Clusters

# Part 1: Getting Dataframe about Postal Code, Borough, and Neighborhoods

### Import Libraries

In [1]:
import numpy as np

import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json

!conda install -c conda-forge geopy
from geopy.geocoders import Nominatim #convert an uaddress into latitude and longitude values

import requests
from pandas.io.json import json_normalize  # transform JSON file into dataframe

import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if necessary
import folium # map rendering library
print("Libraries imported.")

!conda install -c anaconda beautifulsoup4
!conda install -c anaconda lxml
from lxml import etree
from bs4 import BeautifulSoup

Solving environment: ...working... 
  - anaconda/win-64::ca-certificates-2019.8.28-0, anaconda/win-64::certifi-2019.9.11-py37_0, anaconda/win-64::conda-4.8.1-py37_0, anaconda/win-64::openssl-1.1.1d-he774522_2
  - anaconda/win-64::ca-certificates-2019.8.28-0, anaconda/win-64::certifi-2019.9.11-py37_0, anaconda/win-64::openssl-1.1.1d-he774522_2, defaults/win-64::conda-4.8.1-py37_0
  - anaconda/win-64::ca-certificates-2019.8.28-0, anaconda/win-64::certifi-2019.9.11-py37_0, anaconda/win-64::conda-4.8.1-py37_0, defaults/win-64::openssl-1.1.1d-he774522_2
  - anaconda/win-64::ca-certificates-2019.8.28-0, anaconda/win-64::certifi-2019.9.11-py37_0, defaults/win-64::conda-4.8.1-py37_0, defaults/win-64::openssl-1.1.1d-he774522_2
  - anaconda/win-64::certifi-2019.9.11-py37_0, anaconda/win-64::conda-4.8.1-py37_0, defaults/win-64::ca-certificates-2019.8.28-0, defaults/win-64::openssl-1.1.1d-he774522_2
  - anaconda/win-64::certifi-2019.9.11-py37_0, defaults/win-64::ca-certificates-2019.8.28-0, defaul

usage: conda-script.py [-h] [-V] command ...
conda-script.py: error: unrecognized arguments: # uncomment this line if necessary


Solving environment: ...working... done

# All requested packages already installed.

Solving environment: ...working... done

# All requested packages already installed.



### Save the link of postal codes

In [2]:
url_toronto = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [3]:
source = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

## Extract information from the website

### Create BeautifulSoup instance from the website to parse it

In [4]:
soup = BeautifulSoup(source,'lxml')
print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgMonthNamesShort":["","Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"],"wgRequestId":"XiX22QpAADkAAC1sdMUAAAAU","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":935851093,"wgRevisionId":935851093,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Communi

In [5]:
My_table = (soup.find('table',{'class':'wikitable sortable'}))
My_table

<table class="wikitable sortable">
<tbody><tr>
<th>Postcode</th>
<th>Borough</th>
<th>Neighbourhood
</th></tr>
<tr>
<td>M1A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
</td></tr>
<tr>
<td>M4A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
</td></tr>
<tr>
<td>M5A</td>
<td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
<td><a href="/wiki/Regent_Park" title="Regent Park">Harbourfront</a>
</td></tr>
<tr>
<td>M6A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Lawrence_Heights" title="Lawrence Heights">Lawrence Heights</a>
</td></tr>
<tr>
<td>M6A</td>
<td><a href="/wiki/North

In [6]:
table_entry_list = My_table.findAll('td')
table_entry_list 

[<td>M1A</td>, <td>Not assigned</td>, <td>Not assigned
 </td>, <td>M2A</td>, <td>Not assigned</td>, <td>Not assigned
 </td>, <td>M3A</td>, <td><a href="/wiki/North_York" title="North York">North York</a></td>, <td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
 </td>, <td>M4A</td>, <td><a href="/wiki/North_York" title="North York">North York</a></td>, <td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
 </td>, <td>M5A</td>, <td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>, <td><a href="/wiki/Regent_Park" title="Regent Park">Harbourfront</a>
 </td>, <td>M6A</td>, <td><a href="/wiki/North_York" title="North York">North York</a></td>, <td><a href="/wiki/Lawrence_Heights" title="Lawrence Heights">Lawrence Heights</a>
 </td>, <td>M6A</td>, <td><a href="/wiki/North_York" title="North York">North York</a></td>, <td><a href="/wiki/Lawrence_Manor" title="Lawrence Manor">Lawrence Manor</a>
 </td>, <td>M7A</td>, <td>

### Define functions to simplify preprocessing

In [7]:
# This function remove tag from the string
def strip_td(entry_i):
    stringx = str(entry_i)
    stringx = stringx.strip('<td>')
    stringx = stringx.strip('</')
    return stringx

# This function collect title name from the link with a href format
def get_name_from_ahref(stringx):
    try:
        if stringx != 'Not assigned':
            stringx = stringx.split('>')
            stringx = stringx[1]
            stringx = stringx.split('<')
            stringx = stringx[0]  
    except:
        pass
    return stringx

# This function collect title name from oddly named a href format
def get_name_from_residual_list(stringx):
    if type(stringx) == list:
        stringx = stringx[0]
    return stringx

### Extract and save the BeautifulSoup result to three lists

In [8]:
list_PostalCode = []
list_Borough = []
list_Neighborhood = []

for i in range(len(table_entry_list)):
    if i%3 == 0:
        a = strip_td(table_entry_list[i])
        list_PostalCode.append(a)
    elif i%3 == 1:
        b = strip_td(table_entry_list[i])
        b = get_name_from_ahref(b)
        b = get_name_from_residual_list(b)
        list_Borough.append(b)
    else:
        c = strip_td(table_entry_list[i])
        c = get_name_from_ahref(c)
        c = get_name_from_residual_list(c)
        c = c.strip('\n')
        list_Neighborhood.append(c)
         
# some example
print(list_PostalCode[0:10])
print(list_Borough[0:10])
print(list_Neighborhood[0:10])

['M1A', 'M2A', 'M3A', 'M4A', 'M5A', 'M6A', 'M6A', 'M7A', 'M8A', 'M9A']
['Not assigned', 'Not assigned', 'North York', 'North York', 'Downtown Toronto', 'North York', 'North York', 'Downtown Toronto', 'Not assigned', "Queen's Park"]
['Not assigned', 'Not assigned', 'Parkwoods', 'Victoria Village', 'Harbourfront', 'Lawrence Heights', 'Lawrence Manor', "Queen's Park", 'Not assigned', 'Not assigned']


## Combine the list into dataframe

In [9]:
ds_PostalCode = pd.Series(list_PostalCode)
ds_Borough = pd.Series(list_Borough)
ds_Neighborhood = pd.Series(list_Neighborhood)
df = pd.DataFrame(ds_Neighborhood)
df['PostalCode'] = ds_PostalCode
df['Borough'] = ds_Borough
df['Neighborhood'] = ds_Neighborhood
df = df.drop([0], axis='columns')
df

PostalCode           Borough  \
0          M1A      Not assigned   
1          M2A      Not assigned   
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
5          M6A        North York   
6          M6A        North York   
7          M7A  Downtown Toronto   
8          M8A      Not assigned   
9          M9A      Queen's Park   
10         M1B       Scarborough   
11         M1B       Scarborough   
12         M2B      Not assigned   
13         M3B        North York   
14         M4B         East York   
15         M4B         East York   
16         M5B  Downtown Toronto   
17         M5B  Downtown Toronto   
18         M6B        North York   
19         M7B      Not assigned   
20         M8B      Not assigned   
21         M9B         Etobicoke   
22         M9B         Etobicoke   
23         M9B         Etobicoke   
24         M9B         Etobicoke   
25         M9B         Etobicoke   
26         M1C       Scarborough   
27         M1C       Scarborough   
28         M1C       Scarborough   
29         M2C      Not assigned   
30         M3C        North York   
31         M3C        North York   
32         M4C         East York   
33         M5C  Downtown Toronto   
34         M6C              York   
35         M7C      Not assigned   
36         M8C      Not assigned   
37         M9C         Etobicoke   
38         M9C         Etobicoke   
39         M9C         Etobicoke   
40         M9C         Etobicoke   
41         M1E       Scarborough   
42         M1E       Scarborough   
43         M1E       Scarborough   
44         M2E      Not assigned   
45         M3E      Not assigned   
46         M4E      East Toronto   
47         M5E  Downtown Toronto   
48         M6E              York   
49         M7E      Not assigned   
50         M8E      Not assigned   
51         M9E      Not assigned   
52         M1G       Scarborough   
53         M2G      Not assigned   
54         M3G      Not assigned   
55         M4G         East York   
56         M5G  Downtown Toronto   
57         M6G  Downtown Toronto   
58         M7G      Not assigned   
59         M8G      Not assigned   
60         M9G      Not assigned   
61         M1H       Scarborough   
62         M2H        North York   
63         M3H        North York   
64         M3H        North York   
65         M3H        North York   
66         M4H         East York   
67         M5H  Downtown Toronto   
68         M5H  Downtown Toronto   
69         M5H  Downtown Toronto   
70         M6H      West Toronto   
71         M6H      West Toronto   
72         M7H      Not assigned   
73         M8H      Not assigned   
74         M9H      Not assigned   
75         M1J       Scarborough   
76         M2J        North York   
77         M2J        North York   
78         M2J        North York   
79         M3J        North York   
80         M3J        North York   
81         M4J         East York   
82         M5J  Downtown Toronto   
83         M5J  Downtown Toronto   
84         M5J  Downtown Toronto   
85         M6J      West Toronto   
86         M6J      West Toronto   
87         M7J      Not assigned   
88         M8J      Not assigned   
89         M9J      Not assigned   
90         M1K       Scarborough   
91         M1K       Scarborough   
92         M1K       Scarborough   
93         M2K        North York   
94         M3K        North York   
95         M3K        North York   
96         M4K      East Toronto   
97         M4K      East Toronto   
98         M5K  Downtown Toronto   
99         M5K  Downtown Toronto   
100        M6K      West Toronto   
101        M6K      West Toronto   
102        M6K      West Toronto   
103        M7K      Not assigned   
104        M8K      Not assigned   
105        M9K      Not assigned   
106        M1L       Scarborough   
107        M1L       Scarborough   
108        M1L       Scarborough   
109        M2L        North York   
110     

### Only process the cells that have an assigned borough. Ignore cells with a borough that is 'Not assigned'.

In [10]:
df_x = df[df['Borough']!='Not assigned']
df_x = df_x.reset_index()
df_x = df_x.drop('index', axis='columns')
df_x

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M6A        North York   
5          M7A  Downtown Toronto   
6          M9A      Queen's Park   
7          M1B       Scarborough   
8          M1B       Scarborough   
9          M3B        North York   
10         M4B         East York   
11         M4B         East York   
12         M5B  Downtown Toronto   
13         M5B  Downtown Toronto   
14         M6B        North York   
15         M9B         Etobicoke   
16         M9B         Etobicoke   
17         M9B         Etobicoke   
18         M9B         Etobicoke   
19         M9B         Etobicoke   
20         M1C       Scarborough   
21         M1C       Scarborough   
22         M1C       Scarborough   
23         M3C        North York   
24         M3C        North York   
25         M4C         East York   
26         M5C  Downtown Toronto   
27         M6C              York   
28         M9C         Etobicoke   
29         M9C         Etobicoke   
30         M9C         Etobicoke   
31         M9C         Etobicoke   
32         M1E       Scarborough   
33         M1E       Scarborough   
34         M1E       Scarborough   
35         M4E      East Toronto   
36         M5E  Downtown Toronto   
37         M6E              York   
38         M1G       Scarborough   
39         M4G         East York   
40         M5G  Downtown Toronto   
41         M6G  Downtown Toronto   
42         M1H       Scarborough   
43         M2H        North York   
44         M3H        North York   
45         M3H        North York   
46         M3H        North York   
47         M4H         East York   
48         M5H  Downtown Toronto   
49         M5H  Downtown Toronto   
50         M5H  Downtown Toronto   
51         M6H      West Toronto   
52         M6H      West Toronto   
53         M1J       Scarborough   
54         M2J        North York   
55         M2J        North York   
56         M2J        North York   
57         M3J        North York   
58         M3J        North York   
59         M4J         East York   
60         M5J  Downtown Toronto   
61         M5J  Downtown Toronto   
62         M5J  Downtown Toronto   
63         M6J      West Toronto   
64         M6J      West Toronto   
65         M1K       Scarborough   
66         M1K       Scarborough   
67         M1K       Scarborough   
68         M2K        North York   
69         M3K        North York   
70         M3K        North York   
71         M4K      East Toronto   
72         M4K      East Toronto   
73         M5K  Downtown Toronto   
74         M5K  Downtown Toronto   
75         M6K      West Toronto   
76         M6K      West Toronto   
77         M6K      West Toronto   
78         M1L       Scarborough   
79         M1L       Scarborough   
80         M1L       Scarborough   
81         M2L        North York   
82         M2L        North York   
83         M3L        North York   
84         M4L      East Toronto   
85         M4L      East Toronto   
86         M5L  Downtown Toronto   
87         M5L  Downtown Toronto   
88         M6L        North York   
89         M6L        North York   
90         M6L        North York   
91         M9L        North York   
92         M1M       Scarborough   
93         M1M       Scarborough   
94         M1M       Scarborough   
95         M2M        North York   
96         M2M        North York   
97         M3M        North York   
98         M4M      East Toronto   
99         M5M        North York   
100        M5M        North York   
101        M6M              York   
102        M6M              York   
103        M6M              York   
104        M6M              York   
105        M9M        North York   
106        M9M        North York   
107        M1N       Scarborough   
108        M1N       Scarborough   
109        M2N        North York   
110     

### Combine the entry, if they have multiple neighborhoods for the same postal code

The instruction said that:
"More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table."

During the time I retrieve this url: https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M  there is only one entry for M5A. It can be easily verified by searching the word match in a web browser. I also verify it on the dataframe, which postal codes does have multiple entries.

For other rows that have multiple entry for the same postcode, I followed the instruction to combine the neighborhoods as described when they are applicable.

In [11]:
# Investigate postal code number with multiple entry
df_temp = df_x['PostalCode'].value_counts()
print(df_temp.head(20))

print('')
print("We can locate and count the number for each entry")
print('number of M5A entry:', df_temp['M5A'])
print('number of M5R entry:', df_temp['M5R'])
print('number of M5J entry:', df_temp['M5J'], 'and so on')

M9V    8
M8Y    8
M5V    7
M9B    5
M4V    5
M8Z    5
M6M    4
M9R    4
M9C    4
M1V    4
M8X    3
M1E    3
M5R    3
M6L    3
M1L    3
M1M    3
M1C    3
M1K    3
M5T    3
M2J    3
Name: PostalCode, dtype: int64

We can locate and count the number for each entry
number of M5A entry: 1
number of M5R entry: 3
number of M5J entry: 3 and so on


In [12]:
# Re-check df_x
df_x

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M6A        North York   
5          M7A  Downtown Toronto   
6          M9A      Queen's Park   
7          M1B       Scarborough   
8          M1B       Scarborough   
9          M3B        North York   
10         M4B         East York   
11         M4B         East York   
12         M5B  Downtown Toronto   
13         M5B  Downtown Toronto   
14         M6B        North York   
15         M9B         Etobicoke   
16         M9B         Etobicoke   
17         M9B         Etobicoke   
18         M9B         Etobicoke   
19         M9B         Etobicoke   
20         M1C       Scarborough   
21         M1C       Scarborough   
22         M1C       Scarborough   
23         M3C        North York   
24         M3C        North York   
25         M4C         East York   
26         M5C  Downtown Toronto   
27         M6C              York   
28         M9C         Etobicoke   
29         M9C         Etobicoke   
30         M9C         Etobicoke   
31         M9C         Etobicoke   
32         M1E       Scarborough   
33         M1E       Scarborough   
34         M1E       Scarborough   
35         M4E      East Toronto   
36         M5E  Downtown Toronto   
37         M6E              York   
38         M1G       Scarborough   
39         M4G         East York   
40         M5G  Downtown Toronto   
41         M6G  Downtown Toronto   
42         M1H       Scarborough   
43         M2H        North York   
44         M3H        North York   
45         M3H        North York   
46         M3H        North York   
47         M4H         East York   
48         M5H  Downtown Toronto   
49         M5H  Downtown Toronto   
50         M5H  Downtown Toronto   
51         M6H      West Toronto   
52         M6H      West Toronto   
53         M1J       Scarborough   
54         M2J        North York   
55         M2J        North York   
56         M2J        North York   
57         M3J        North York   
58         M3J        North York   
59         M4J         East York   
60         M5J  Downtown Toronto   
61         M5J  Downtown Toronto   
62         M5J  Downtown Toronto   
63         M6J      West Toronto   
64         M6J      West Toronto   
65         M1K       Scarborough   
66         M1K       Scarborough   
67         M1K       Scarborough   
68         M2K        North York   
69         M3K        North York   
70         M3K        North York   
71         M4K      East Toronto   
72         M4K      East Toronto   
73         M5K  Downtown Toronto   
74         M5K  Downtown Toronto   
75         M6K      West Toronto   
76         M6K      West Toronto   
77         M6K      West Toronto   
78         M1L       Scarborough   
79         M1L       Scarborough   
80         M1L       Scarborough   
81         M2L        North York   
82         M2L        North York   
83         M3L        North York   
84         M4L      East Toronto   
85         M4L      East Toronto   
86         M5L  Downtown Toronto   
87         M5L  Downtown Toronto   
88         M6L        North York   
89         M6L        North York   
90         M6L        North York   
91         M9L        North York   
92         M1M       Scarborough   
93         M1M       Scarborough   
94         M1M       Scarborough   
95         M2M        North York   
96         M2M        North York   
97         M3M        North York   
98         M4M      East Toronto   
99         M5M        North York   
100        M5M        North York   
101        M6M              York   
102        M6M              York   
103        M6M              York   
104        M6M              York   
105        M9M        North York   
106        M9M        North York   
107        M1N       Scarborough   
108        M1N       Scarborough   
109        M2N        North York   
110     

In [13]:
df_x['Neighborhood'] = df_x['Neighborhood']+", "
df_x

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M6A        North York   
5          M7A  Downtown Toronto   
6          M9A      Queen's Park   
7          M1B       Scarborough   
8          M1B       Scarborough   
9          M3B        North York   
10         M4B         East York   
11         M4B         East York   
12         M5B  Downtown Toronto   
13         M5B  Downtown Toronto   
14         M6B        North York   
15         M9B         Etobicoke   
16         M9B         Etobicoke   
17         M9B         Etobicoke   
18         M9B         Etobicoke   
19         M9B         Etobicoke   
20         M1C       Scarborough   
21         M1C       Scarborough   
22         M1C       Scarborough   
23         M3C        North York   
24         M3C        North York   
25         M4C         East York   
26         M5C  Downtown Toronto   
27         M6C              York   
28         M9C         Etobicoke   
29         M9C         Etobicoke   
30         M9C         Etobicoke   
31         M9C         Etobicoke   
32         M1E       Scarborough   
33         M1E       Scarborough   
34         M1E       Scarborough   
35         M4E      East Toronto   
36         M5E  Downtown Toronto   
37         M6E              York   
38         M1G       Scarborough   
39         M4G         East York   
40         M5G  Downtown Toronto   
41         M6G  Downtown Toronto   
42         M1H       Scarborough   
43         M2H        North York   
44         M3H        North York   
45         M3H        North York   
46         M3H        North York   
47         M4H         East York   
48         M5H  Downtown Toronto   
49         M5H  Downtown Toronto   
50         M5H  Downtown Toronto   
51         M6H      West Toronto   
52         M6H      West Toronto   
53         M1J       Scarborough   
54         M2J        North York   
55         M2J        North York   
56         M2J        North York   
57         M3J        North York   
58         M3J        North York   
59         M4J         East York   
60         M5J  Downtown Toronto   
61         M5J  Downtown Toronto   
62         M5J  Downtown Toronto   
63         M6J      West Toronto   
64         M6J      West Toronto   
65         M1K       Scarborough   
66         M1K       Scarborough   
67         M1K       Scarborough   
68         M2K        North York   
69         M3K        North York   
70         M3K        North York   
71         M4K      East Toronto   
72         M4K      East Toronto   
73         M5K  Downtown Toronto   
74         M5K  Downtown Toronto   
75         M6K      West Toronto   
76         M6K      West Toronto   
77         M6K      West Toronto   
78         M1L       Scarborough   
79         M1L       Scarborough   
80         M1L       Scarborough   
81         M2L        North York   
82         M2L        North York   
83         M3L        North York   
84         M4L      East Toronto   
85         M4L      East Toronto   
86         M5L  Downtown Toronto   
87         M5L  Downtown Toronto   
88         M6L        North York   
89         M6L        North York   
90         M6L        North York   
91         M9L        North York   
92         M1M       Scarborough   
93         M1M       Scarborough   
94         M1M       Scarborough   
95         M2M        North York   
96         M2M        North York   
97         M3M        North York   
98         M4M      East Toronto   
99         M5M        North York   
100        M5M        North York   
101        M6M              York   
102        M6M              York   
103        M6M              York   
104        M6M              York   
105        M9M        North York   
106        M9M        North York   
107        M1N       Scarborough   
108        M1N       Scarborough   
109        M2N        North York   
110     

In [14]:
df_x[df_x['Neighborhood'] == 'Not assigned, '].sum()

PostalCode                 M9A
Borough           Queen's Park
Neighborhood    Not assigned, 
dtype: object

In [15]:
# Combine several neighborhood with the .sum() function, then reset index.
df_y = df_x.groupby(['PostalCode','Borough']).sum().reset_index()
df_y

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
30         M3K        North York   
31         M3L        North York   
32         M3M        North York   
33         M3N        North York   
34         M4A        North York   
35         M4B         East York   
36         M4C         East York   
37         M4E      East Toronto   
38         M4G         East York   
39         M4H         East York   
40         M4J         East York   
41         M4K      East Toronto   
42         M4L      East Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
45         M4P   Central Toronto   
46         M4R   Central Toronto   
47         M4S   Central Toronto   
48         M4T   Central Toronto   
49         M4V   Central Toronto   
50         M4W  Downtown Toronto   
51         M4X  Downtown Toronto   
52         M4Y  Downtown Toronto   
53         M5A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
58         M5H  Downtown Toronto   
59         M5J  Downtown Toronto   
60         M5K  Downtown Toronto   
61         M5L  Downtown Toronto   
62         M5M        North York   
63         M5N   Central Toronto   
64         M5P   Central Toronto   
65         M5R   Central Toronto   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
71         M6A        North York   
72         M6B        North York   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A  Downtown Toronto   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A      Queen's Park   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood  
0                                     Rouge, Malvern,   
1             Highland Creek, Rouge Hill, Port Union,   
2                  Guildwood, Morningside, West Hill,   
3                              

In [16]:
# remove last comma
def my_cut_last_char_func(stringx):
    stringx = stringx[0:-2]
    return stringx

df_y['Neighborhood'] = df_y['Neighborhood'].apply(my_cut_last_char_func) 
df_y

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
30         M3K        North York   
31         M3L        North York   
32         M3M        North York   
33         M3N        North York   
34         M4A        North York   
35         M4B         East York   
36         M4C         East York   
37         M4E      East Toronto   
38         M4G         East York   
39         M4H         East York   
40         M4J         East York   
41         M4K      East Toronto   
42         M4L      East Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
45         M4P   Central Toronto   
46         M4R   Central Toronto   
47         M4S   Central Toronto   
48         M4T   Central Toronto   
49         M4V   Central Toronto   
50         M4W  Downtown Toronto   
51         M4X  Downtown Toronto   
52         M4Y  Downtown Toronto   
53         M5A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
58         M5H  Downtown Toronto   
59         M5J  Downtown Toronto   
60         M5K  Downtown Toronto   
61         M5L  Downtown Toronto   
62         M5M        North York   
63         M5N   Central Toronto   
64         M5P   Central Toronto   
65         M5R   Central Toronto   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
71         M6A        North York   
72         M6B        North York   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A  Downtown Toronto   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A      Queen's Park   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood  
0                                       Rouge, Malvern  
1               Highland Creek, Rouge Hill, Port Union  
2                    Guildwood, Morningside, West Hill  
3                              

### If a cell has a borough but a 'Not assigned' neighborhood, then the neighborhood will be the same as the borough. 

If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.

In [17]:
df_y[df_y['Neighborhood'] == 'Not assigned']

PostalCode       Borough  Neighborhood
93        M9A  Queen's Park  Not assigned

In [18]:
# Since there is only one, we will simply assign it:
df_y.loc[df_y['PostalCode']=='M9A','Neighborhood'] = df_y.loc[df_y['PostalCode']=='M9A','Borough']
df_y[df_y['PostalCode'] == 'M9A']

PostalCode       Borough  Neighborhood
93        M9A  Queen's Park  Queen's Park

### Cleaning up the dataframe:

In [19]:
# Function below help to show the same entry as expected from the notebook
def postal_code_example(stringx):
    example_of_postal_code = {'M5G':1, 'M2H':2, 'M4B':3, 'M1J':4, 'M4G':5, 'M4M':6, 'M1R':7, 'M9V':8, 'M9L':9, 
                          'M5V':10, 'M1B':11, 'M5A':12}
    try:
        rank = example_of_postal_code[stringx]
    except:
        rank = 13
    return rank

# We use temporary 'rank' column to help sort the dataframe
df_y['rank'] = df_y['PostalCode'].map(postal_code_example)
df_y = df_y.sort_values('rank')
df_y = df_y.drop('rank', axis = 'columns')
df_y = df_y.reset_index()
df_y = df_y[['PostalCode', 'Borough', 'Neighborhood']]
df_y.head(12)

PostalCode           Borough  \
0         M5G  Downtown Toronto   
1         M2H        North York   
2         M4B         East York   
3         M1J       Scarborough   
4         M4G         East York   
5         M4M      East Toronto   
6         M1R       Scarborough   
7         M9V         Etobicoke   
8         M9L        North York   
9         M5V  Downtown Toronto   
10        M1B       Scarborough   
11        M5A  Downtown Toronto   

                                         Neighborhood  
0                                  Central Bay Street  
1                                   Hillcrest Village  
2                     Woodbine Gardens, Parkview Hill  
3                                 Scarborough Village  
4                                             Leaside  
5                                     Studio District  
6                                   Maryvale, Wexford  
7   Albion Gardens, Beaumond Heights, Humbergate, ...  
8                                       Humber Summit  
9   CN Tower, Bathurst Quay, Island airport, Harbo...  
10                                     Rouge, Malvern  
11                                       Harbourfront

In [20]:
# save to final dataframe
df_final = df_y

# Final Dataframe

So here is the final dataframe:

In [21]:
print(df_final.shape)

(103, 3)


Hence, our final dataframe has 103 rows and 3 columns.

Some of the example of the top rows are shown in the following cell:

In [22]:
df_final.head(14)

PostalCode           Borough  \
0         M5G  Downtown Toronto   
1         M2H        North York   
2         M4B         East York   
3         M1J       Scarborough   
4         M4G         East York   
5         M4M      East Toronto   
6         M1R       Scarborough   
7         M9V         Etobicoke   
8         M9L        North York   
9         M5V  Downtown Toronto   
10        M1B       Scarborough   
11        M5A  Downtown Toronto   
12        M5T  Downtown Toronto   
13        M6A        North York   

                                         Neighborhood  
0                                  Central Bay Street  
1                                   Hillcrest Village  
2                     Woodbine Gardens, Parkview Hill  
3                                 Scarborough Village  
4                                             Leaside  
5                                     Studio District  
6                                   Maryvale, Wexford  
7   Albion Gardens, Beaumond Heights, Humbergate, ...  
8                                       Humber Summit  
9   CN Tower, Bathurst Quay, Island airport, Harbo...  
10                                     Rouge, Malvern  
11                                       Harbourfront  
12          Chinatown, Grange Park, Kensington Market  
13                   Lawrence Heights, Lawrence Manor

In [23]:
#
#
#
#
#

# Part 2: Getting the Dataframe with Coordinates

### Load the coordinates file

here is a link to a csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data

use pandas to read the csv file from the following url:

In [24]:
coordinate_list = pd.read_csv('https://cocl.us/Geospatial_data')

In [25]:
coordinate_list

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
30          M3K  43.737473 -79.464763
31          M3L  43.739015 -79.506944
32          M3M  43.728496 -79.495697
33          M3N  43.761631 -79.520999
34          M4A  43.725882 -79.315572
35          M4B  43.706397 -79.309937
36          M4C  43.695344 -79.318389
37          M4E  43.676357 -79.293031
38          M4G  43.709060 -79.363452
39          M4H  43.705369 -79.349372
40          M4J  43.685347 -79.338106
41          M4K  43.679557 -79.352188
42          M4L  43.668999 -79.315572
43          M4M  43.659526 -79.340923
44          M4N  43.728020 -79.388790
45          M4P  43.712751 -79.390197
46          M4R  43.715383 -79.405678
47          M4S  43.704324 -79.388790
48          M4T  43.689574 -79.383160
49          M4V  43.686412 -79.400049
50          M4W  43.679563 -79.377529
51          M4X  43.667967 -79.367675
52          M4Y  43.665860 -79.383160
53          M5A  43.654260 -79.360636
54          M5B  43.657162 -79.378937
55          M5C  43.651494 -79.375418
56          M5E  43.644771 -79.373306
57          M5G  43.657952 -79.387383
58          M5H  43.650571 -79.384568
59          M5J  43.640816 -79.381752
60          M5K  43.647177 -79.381576
61          M5L  43.648198 -79.379817
62          M5M  43.733283 -79.419750
63          M5N  43.711695 -79.416936
64          M5P  43.696948 -79.411307
65          M5R  43.672710 -79.405678
66          M5S  43.662696 -79.400049
67          M5T  43.653206 -79.400049
68          M5V  43.628947 -79.394420
69          M5W  43.646435 -79.374846
70          M5X  43.648429 -79.382280
71          M6A  43.718518 -79.464763
72          M6B  43.709577 -79.445073
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

### Matching the postal code, to get the latitude and longitude

We need to match and input this list of postal code latitude, longitude, and then merge it to our Toronto, Neighborhood Dataframe.

For reference, that dataframe looks like this:

In [26]:
df_final

PostalCode           Borough  \
0          M5G  Downtown Toronto   
1          M2H        North York   
2          M4B         East York   
3          M1J       Scarborough   
4          M4G         East York   
5          M4M      East Toronto   
6          M1R       Scarborough   
7          M9V         Etobicoke   
8          M9L        North York   
9          M5V  Downtown Toronto   
10         M1B       Scarborough   
11         M5A  Downtown Toronto   
12         M5T  Downtown Toronto   
13         M6A        North York   
14         M6B        North York   
15         M6C              York   
16         M5W  Downtown Toronto   
17         M5X  Downtown Toronto   
18         M5P   Central Toronto   
19         M5R   Central Toronto   
20         M5N   Central Toronto   
21         M5M        North York   
22         M5L  Downtown Toronto   
23         M5K  Downtown Toronto   
24         M5J  Downtown Toronto   
25         M5H  Downtown Toronto   
26         M5E  Downtown Toronto   
27         M5S  Downtown Toronto   
28         M6E              York   
29         M6J      West Toronto   
30         M6H      West Toronto   
31         M9R         Etobicoke   
32         M9P         Etobicoke   
33         M9N              York   
34         M9M        North York   
35         M9C         Etobicoke   
36         M9B         Etobicoke   
37         M9A      Queen's Park   
38         M8Z         Etobicoke   
39         M8Y         Etobicoke   
40         M8X         Etobicoke   
41         M6G  Downtown Toronto   
42         M8W         Etobicoke   
43         M7Y      East Toronto   
44         M7R       Mississauga   
45         M7A  Downtown Toronto   
46         M6S      West Toronto   
47         M6R      West Toronto   
48         M6P      West Toronto   
49         M6N              York   
50         M6M              York   
51         M6L        North York   
52         M6K      West Toronto   
53         M8V         Etobicoke   
54         M5C  Downtown Toronto   
55         M4X  Downtown Toronto   
56         M4Y  Downtown Toronto   
57         M2P        North York   
58         M2N        North York   
59         M2M        North York   
60         M2L        North York   
61         M2K        North York   
62         M2J        North York   
63         M1X       Scarborough   
64         M1W       Scarborough   
65         M1V       Scarborough   
66         M2R        North York   
67         M1T       Scarborough   
68         M1P       Scarborough   
69         M1N       Scarborough   
70         M1M       Scarborough   
71         M1L       Scarborough   
72         M1K       Scarborough   
73         M1H       Scarborough   
74         M1G       Scarborough   
75         M1E       Scarborough   
76         M1C       Scarborough   
77         M1S       Scarborough   
78         M5B  Downtown Toronto   
79         M3A        North York   
80         M3C        North York   
81         M4W  Downtown Toronto   
82         M4V   Central Toronto   
83         M4T   Central Toronto   
84         M4S   Central Toronto   
85         M4R   Central Toronto   
86         M4P   Central Toronto   
87         M4N   Central Toronto   
88         M4L      East Toronto   
89         M4K      East Toronto   
90         M3B        North York   
91         M4J         East York   
92         M4E      East Toronto   
93         M4C         East York   
94         M4A        North York   
95         M3N        North York   
96         M3M        North York   
97         M3L        North York   
98         M3K        North York   
99         M3J        North York   
100        M3H        North York   
101        M4H         East York   
102        M9W         Etobicoke   

                                          Neighborhood  
0                                   Central Bay Street  
1                                    Hillcrest Village  
2                      Woodbine Gardens, Parkview Hill  
3                              

In [27]:
rowscount, columnscount = df_final.shape
rowscount

103

In [28]:
df_final['Latitude'] = 0
df_final['Longitude'] = 0

for index in range(rowscount):
    postal_code = df_final['PostalCode'].iloc[index]
    dtemp = coordinate_list[(coordinate_list['Postal Code']==postal_code)]
    lat = dtemp['Latitude'].values[0]
    lon = dtemp['Longitude'].values[0]
    df_final['Latitude'].iloc[index] = lat
    df_final['Longitude'].iloc[index] = lon
    print(postal_code, lat, lon)

M5G 43.6579524 -79.3873826
M2H 43.8037622 -79.3634517
M4B 43.7063972 -79.309937
M1J 43.7447342 -79.23947609999999
M4G 43.7090604 -79.3634517
M4M 43.6595255 -79.340923
M1R 43.750071500000004 -79.2958491
M9V 43.739416399999996 -79.5884369
M9L 43.7563033 -79.56596329999999
M5V 43.6289467 -79.3944199
M1B 43.806686299999996 -79.19435340000001
M5A 43.6542599 -79.3606359
M5T 43.6532057 -79.4000493
M6A 43.718517999999996 -79.46476329999999
M6B 43.709577 -79.44507259999999
M6C 43.6937813 -79.42819140000002
M5W 43.6464352 -79.37484599999999
M5X 43.6484292 -79.3822802
M5P 43.6969476 -79.41130720000001
M5R 43.6727097 -79.40567840000001
M5N 43.7116948 -79.41693559999999
M5M 43.7332825 -79.4197497
M5L 43.6481985 -79.37981690000001
M5K 43.6471768 -79.38157640000001
M5J 43.6408157 -79.38175229999999
M5H 43.65057120000001 -79.3845675
M5E 43.644770799999996 -79.3733064
M5S 43.6626956 -79.4000493
M6E 43.6890256 -79.453512
M6J 43.647926700000006 -79.4197497
M6H 43.66900510000001 -79.4422593
M9R 43.6889054

C:\Users\aribimop\Anaconda3\lib\site-packages\pandas\core\indexing.py:205: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_with_indexer(indexer, value)


### Dataframe with Latitude and Longitude

In [29]:
df_final.head(14)

PostalCode           Borough  \
0         M5G  Downtown Toronto   
1         M2H        North York   
2         M4B         East York   
3         M1J       Scarborough   
4         M4G         East York   
5         M4M      East Toronto   
6         M1R       Scarborough   
7         M9V         Etobicoke   
8         M9L        North York   
9         M5V  Downtown Toronto   
10        M1B       Scarborough   
11        M5A  Downtown Toronto   
12        M5T  Downtown Toronto   
13        M6A        North York   

                                         Neighborhood   Latitude  Longitude  
0                                  Central Bay Street  43.657952 -79.387383  
1                                   Hillcrest Village  43.803762 -79.363452  
2                     Woodbine Gardens, Parkview Hill  43.706397 -79.309937  
3                                 Scarborough Village  43.744734 -79.239476  
4                                             Leaside  43.709060 -79.363452  
5                                     Studio District  43.659526 -79.340923  
6                                   Maryvale, Wexford  43.750072 -79.295849  
7   Albion Gardens, Beaumond Heights, Humbergate, ...  43.739416 -79.588437  
8                                       Humber Summit  43.756303 -79.565963  
9   CN Tower, Bathurst Quay, Island airport, Harbo...  43.628947 -79.394420  
10                                     Rouge, Malvern  43.806686 -79.194353  
11                                       Harbourfront  43.654260 -79.360636  
12          Chinatown, Grange Park, Kensington Market  43.653206 -79.400049  
13                   Lawrence Heights, Lawrence Manor  43.718518 -79.464763

# Part 3: Neigborhood Clustering

### Render Toronto map and each postal code

In [30]:
import folium # map rendering library

In [31]:
# Get one coordinate for entire Toronto. 
latitude = 43.6532
longitude = -79.3832
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6532, -79.3832.


In [32]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_final['Latitude'], df_final['Longitude'], df_final['Borough'], df_final['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## Cluster based on venues of interest

### Explore first neighborhood

In [33]:
CLIENT_ID = 'PFYGTWGIW5MKCU2LBRZYC1A1JLZ43ULPO4FYHYCRMJUB3GID' # your Foursquare ID
CLIENT_SECRET = 'I05LS14HD54RGZ5LYLJ3JZ5TIDP5S2PKTVIKG4QLJ1Z2M3MX' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

# print('Your credentails:')
# print('CLIENT_ID: ' + CLIENT_ID)
# print('CLIENT_SECRET:' + CLIENT_SECRET)

In [34]:
df_final.loc[0, 'Neighborhood']

'Central Bay Street'

In [35]:
lat = df_final.loc[0, 'Latitude']
lon = df_final.loc[0, 'Longitude']

In [36]:
# Get 100 nearest venues
LIMIT = 100   #limit number of venues

radius = 500  #define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    lat, 
    lon, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=PFYGTWGIW5MKCU2LBRZYC1A1JLZ43ULPO4FYHYCRMJUB3GID&client_secret=I05LS14HD54RGZ5LYLJ3JZ5TIDP5S2PKTVIKG4QLJ1Z2M3MX&v=20180605&ll=43.6579524,-79.3873826&radius=500&limit=100'

In [37]:
# import requests # library to handle requests
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5e297e7ddd0f85001c73de6a'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Bay Street Corridor',
  'headerFullLocation': 'Bay Street Corridor, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 85,
  'suggestedBounds': {'ne': {'lat': 43.6624524045, 'lng': -79.38117421839567},
   'sw': {'lat': 43.6534523955, 'lng': -79.39359098160432}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '537d4d6d498ec171ba22e7fe',
       'name': "Jimmy's Coffee",
       'location': {'address': '82 Gerrard Street W',
        'crossStreet': 'Gerrard & LaPlante',
        'lat': 43.65842123574496,
        'lng': -79.38561319551111,
        'label

In [38]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [39]:
venues = results['response']['groups'][0]['items']

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name                  categories        lat  \
0                     Jimmy's Coffee                 Coffee Shop  43.658421   
1                        Tim Hortons                 Coffee Shop  43.658570   
2                      Hailed Coffee                 Coffee Shop  43.658833   
3            The Elm Tree Restaurant  Modern European Restaurant  43.657397   
4  The Queen and Beaver Public House                   Gastropub  43.657472   

         lng  
0 -79.385613  
1 -79.385123  
2 -79.383684  
3 -79.383761  
4 -79.383524

In [40]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

85 venues were returned by Foursquare.


### Explore all neighborhoods

In [41]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### Get data of venues in Toronto

In [43]:
toronto_venues = getNearbyVenues(names=df_final['Neighborhood'],
                                   latitudes=df_final['Latitude'],
                                   longitudes=df_final['Longitude']
                                  )

Central Bay Street
Hillcrest Village
Woodbine Gardens, Parkview Hill
Scarborough Village
Leaside
Studio District
Maryvale, Wexford
Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown
Humber Summit
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Rouge, Malvern
Harbourfront
Chinatown, Grange Park, Kensington Market
Lawrence Heights, Lawrence Manor
Glencairn
Humewood-Cedarvale
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Forest Hill North, Forest Hill West
The Annex, North Midtown, Yorkville
Roselawn
Bedford Park, Lawrence Manor East
Commerce Court, Victoria Hotel
Design Exchange, Toronto Dominion Centre
Harbourfront East, Toronto Islands, Union Station
Adelaide, King, Richmond
Berczy Park
Harbord, University of Toronto
Caledonia-Fairbanks
Little Portugal, Trinity
Dovercourt Village, Dufferin
Kingsview Village, Martin Grove Gardens, Richview Garden

In [44]:
print(toronto_venues.shape)
toronto_venues.head()

(2220, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Central Bay Street              43.657952              -79.387383   
1  Central Bay Street              43.657952              -79.387383   
2  Central Bay Street              43.657952              -79.387383   
3  Central Bay Street              43.657952              -79.387383   
4  Central Bay Street              43.657952              -79.387383   

                               Venue  Venue Latitude  Venue Longitude  \
0                     Jimmy's Coffee       43.658421       -79.385613   
1                        Tim Hortons       43.658570       -79.385123   
2                      Hailed Coffee       43.658833       -79.383684   
3            The Elm Tree Restaurant       43.657397       -79.383761   
4  The Queen and Beaver Public House       43.657472       -79.383524   

               Venue Category  
0                 Coffee Shop  
1                 Coffee Shop  
2                 Coffee Shop  
3  Modern European Restaurant  
4                   Gastropub

### Check the number of venues for each neighborhood

In [45]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Agincourt                                                               5   
Agincourt North, L'Amoreaux East, Milliken, Ste...                      2   
Albion Gardens, Beaumond Heights, Humbergate, J...                      9   
Alderwood, Long Branch                                                 10   
Bathurst Manor, Downsview North, Wilson Heights                        19   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      25   
Berczy Park                                                            58   
Birch Cliff, Cliffside West                                             4   
Bloordale Gardens, Eringate, Markland Wood, Old...                      7   
Brockton, Exhibition Place, Parkdale Village                           24   
Business Reply Mail Processing Centre 969 Eastern                      18   
CFB Toronto, Downsview East                                             3   
CN Tower, Bathurst Quay, Island airport, Harbou...                     14   
Cabbagetown, St. James Town                                            48   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  11   
Cedarbrae                                                               8   
Central Bay Street                                                     85   
Chinatown, Grange Park, Kensington Market                              86   
Christie                                                               18   
Church and Wellesley                                                   85   
Clairlea, Golden Mile, Oakridge                                        10   
Clarks Corners, Sullivan, Tam O'Shanter                                13   
Cliffcrest, Cliffside, Scarborough Village West                         3   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             38   
Davisville North                                                        8   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     14   
Del Ray, Keelesdale, Mount Dennis, Silverthorn                          5   
Design Exchange, Toronto Dominion Centre                              100   
Don Mills North                                                         6   
Dorset Park, Scarborough Town Centre, Wexford H...                      5   
Dovercourt Village, Dufferin                                           16   
Downsview Central                                                       4   
Downsview Northwest                                                     4   
Downsview West                                                          4   
Downsview, North Park, Upwood Park                                      4   
East Birchmount Park, Ionview, Kennedy Park                             5   
East Toronto                                                            3   
Emery, Humberlea                                                        1   
Fairview, Henry Farm, Oriole                                           61   
First Canadian Place, Underground city                                100   
Flemingdon Park, Don Mills South                                       23   
Forest Hill North, Forest Hill West                                     4   
Glencairn                                                               4   
Guildwood, Morningside, West Hill                                       7   
Harbord, University of Toronto                                         35   
Harbourfront                                                           46   
Harbourfront East, Toronto Islands, Union Statio

In [46]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 266 uniques categories.


### Analyze each neighborhood

In [47]:
#toronto_venues.head()

In [48]:
# del toronto_onehot

In [49]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
#toronto_onehot.head()

In [50]:
# add neighborhood column back to dataframe
df_neigbor = toronto_venues['Neighborhood']

toronto_onehot.drop(labels=['Neighborhood'], axis=1,inplace = True)

toronto_onehot.insert(0, 'Neighborhood', df_neigbor)
toronto_onehot.head()

Neighborhood  Accessories Store  Afghan Restaurant  Airport  \
0  Central Bay Street                  0                  0        0   
1  Central Bay Street                  0                  0        0   
2  Central Bay Street                  0                  0        0   
3  Central Bay Street                  0                  0        0   
4  Central Bay Street                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0               0                 0                 0                   0   
1               0                 0                 0                   0   
2               0                 0                 0                   0   
3               0                 0                 0                   0   
4               0                 0                 0                   0   

   Beach  Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  \
0      0                0         0           0                   0   
1      0                0         0           0                   0   
2      0                0         0           0                   0   
3      0                0         0           0                   0   
4      0                0         0           0                   0   

   Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0          0       0              0          0         0   
1          0       0              0          0         0   
2          0       0              0          0         0   
3          0       0              0          0         0   
4          0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0                0         0       

In [51]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Accessories Store  \
0                            Adelaide, King, Richmond           0.000000   
1                                           Agincourt           0.000000   
2   Agincourt North, L'Amoreaux East, Milliken, St...           0.000000   
3   Albion Gardens, Beaumond Heights, Humbergate, ...           0.000000   
4                              Alderwood, Long Branch           0.000000   
5     Bathurst Manor, Downsview North, Wilson Heights           0.000000   
6                                     Bayview Village           0.000000   
7                   Bedford Park, Lawrence Manor East           0.000000   
8                                         Berczy Park           0.000000   
9                         Birch Cliff, Cliffside West           0.000000   
10  Bloordale Gardens, Eringate, Markland Wood, Ol...           0.000000   
11       Brockton, Exhibition Place, Parkdale Village           0.000000   
12  Business Reply Mail Processing Centre 969 Eastern           0.000000   
13                        CFB Toronto, Downsview East           0.000000   
14  CN Tower, Bathurst Quay, Island airport, Harbo...           0.000000   
15                        Cabbagetown, St. James Town           0.000000   
16                                Caledonia-Fairbanks           0.000000   
17              Canada Post Gateway Processing Centre           0.000000   
18                                          Cedarbrae           0.000000   
19                                 Central Bay Street           0.000000   
20          Chinatown, Grange Park, Kensington Market           0.000000   
21                                           Christie           0.000000   
22                               Church and Wellesley           0.000000   
23                    Clairlea, Golden Mile, Oakridge           0.000000   
24            Clarks Corners, Sullivan, Tam O'Shanter           0.000000   
25    Cliffcrest, Cliffside, Scarborough Village West           0.000000   
26                     Commerce Court, Victoria Hotel           0.000000   
27                                         Davisville           0.000000   
28                                   Davisville North           0.000000   
29  Deer Park, Forest Hill SE, Rathnelly, South Hi...           0.000000   
30     Del Ray, Keelesdale, Mount Dennis, Silverthorn           0.000000   
31           Design Exchange, Toronto Dominion Centre           0.000000   
32                                    Don Mills North           0.000000   
33  Dorset Park, Scarborough Town Centre, Wexford ...           0.000000   
34                       Dovercourt Village, Dufferin           0.000000   
35                                  Downsview Central           0.000000   
36                                Downsview Northwest           0.000000   
37                                     Downsview West           0.000000   
38                 Downsview, North Park, Upwood Park           0.000000   
39        East Birchmount Park, Ionview, Kennedy Park           0.000000   
40                                       East Toronto           0.000000   
41                                   Emery, Humberlea           0.000000   
42                       Fairview, Henry Farm, Oriole           0.000000   
43             First Canadian Place, Underground city           0.000000   
44                   Flemingdon Park, Don Mills South           0.000000   
45                Forest Hill North, Forest Hill West           0.000000   
46                                          Glencairn           0.000000   
47                  Guildwood, Morningside, West Hill           0.000000   
48                     Harbord, University of Toronto           0.000000   
49                                       Harbourfront           0.000000   
50  Harbourfront East, Toronto Islands, Union Station           0.000000   
51                      High Park, The Junction South           0.000000   
52           

In [52]:
toronto_grouped.shape

(99, 266)

In [53]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
            venue  freq
0     Coffee Shop  0.07
1            Café  0.04
2      Steakhouse  0.04
3             Bar  0.04
4  Cosmetics Shop  0.03


----Agincourt----
                       venue  freq
0             Clothing Store   0.2
1                     Lounge   0.2
2             Breakfast Spot   0.2
3  Latin American Restaurant   0.2
4               Skating Rink   0.2


----Agincourt North, L'Amoreaux East, Milliken, Steeles East----
                             venue  freq
0                       Playground   0.5
1                             Park   0.5
2                    Metro Station   0.0
3  Molecular Gastronomy Restaurant   0.0
4       Modern European Restaurant   0.0


----Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
                  venue  freq
0         Grocery Store  0.22
1           Pizza Place  0.11
2   Fried Chicken Joint  0.11
3        Sandwich Place  0.11
4  Fast Fo

               venue  freq
0        Flower Shop  0.07
1     Discount Store  0.07
2       Burger Joint  0.07
3  Convenience Store  0.07
4      Burrito Place  0.07


----L'Amoreaux West----
                  venue  freq
0  Fast Food Restaurant  0.20
1    Chinese Restaurant  0.13
2           Coffee Shop  0.07
3       Bubble Tea Shop  0.07
4           Supermarket  0.07


----Lawrence Heights, Lawrence Manor----
                venue  freq
0      Clothing Store  0.18
1   Accessories Store  0.09
2         Coffee Shop  0.09
3         Event Space  0.09
4  Miscellaneous Shop  0.09


----Lawrence Park----
                venue  freq
0                Park   0.2
1  Dim Sum Restaurant   0.2
2                Lake   0.2
3            Bus Line   0.2
4         Swim School   0.2


----Leaside----
                    venue  freq
0             Coffee Shop  0.12
1     Sporting Goods Shop  0.09
2            Burger Joint  0.06
3  Furniture / Home Store  0.06
4               Bike Shop  0.03


----Little Portug

In [54]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [55]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  \
0                           Adelaide, King, Richmond   
1                                          Agincourt   
2  Agincourt North, L'Amoreaux East, Milliken, St...   
3  Albion Gardens, Beaumond Heights, Humbergate, ...   
4                             Alderwood, Long Branch   

       1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                Coffee Shop            Steakhouse                  Café   
1  Latin American Restaurant                Lounge        Breakfast Spot   
2                       Park            Playground           Yoga Studio   
3              Grocery Store        Sandwich Place          Liquor Store   
4                Pizza Place           Coffee Shop                   Pub   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                   Bar                 Hotel            Restaurant   
1          Skating Rink        Clothing Store   Dumpling Restaurant   
2            Donut Shop    Dim Sum Restaurant                 Diner   
3  Fast Food Restaurant            Beer Store   Fried Chicken Joint   
4                   Gym          Dance Studio              Pharmacy   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0       Thai Restaurant      Asian Restaurant        Cosmetics Shop   
1               Dog Run      Doner Restaurant            Donut Shop   
2        Discount Store               Dog Run      Doner Restaurant   
3              Pharmacy           Pizza Place          Concert Hall   
4          Skating Rink        Sandwich Place    Athletics & Sports   

       10th Most Common Venue  
0              Breakfast Spot  
1                   Drugstore  
2         Dumpling Restaurant  
3  Construction & Landscaping  
4          Dim Sum Restaurant

## Use machine learning to cluster the neighborhood

In [56]:
# set number of clusters
kclusters = 8

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 3, 0, 0, 0, 0, 0, 0, 0])

In [57]:
# df_final = toronto_data

# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_final

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')


In [58]:
toronto_merged['Cluster Labels'].value_counts()
toronto_merged = toronto_merged.dropna()
toronto_merged.head(10)

PostalCode           Borough  \
0        M5G  Downtown Toronto   
1        M2H        North York   
2        M4B         East York   
3        M1J       Scarborough   
4        M4G         East York   
5        M4M      East Toronto   
6        M1R       Scarborough   
7        M9V         Etobicoke   
8        M9L        North York   
9        M5V  Downtown Toronto   

                                        Neighborhood   Latitude  Longitude  \
0                                 Central Bay Street  43.657952 -79.387383   
1                                  Hillcrest Village  43.803762 -79.363452   
2                    Woodbine Gardens, Parkview Hill  43.706397 -79.309937   
3                                Scarborough Village  43.744734 -79.239476   
4                                            Leaside  43.709060 -79.363452   
5                                    Studio District  43.659526 -79.340923   
6                                  Maryvale, Wexford  43.750072 -79.295849   
7  Albion Gardens, Beaumond Heights, Humbergate, ...  43.739416 -79.588437   
8                                      Humber Summit  43.756303 -79.565963   
9  CN Tower, Bathurst Quay, Island airport, Harbo...  43.628947 -79.394420   

   Cluster Labels      1st Most Common Venue 2nd Most Common Venue  \
0             0.0                Coffee Shop        Sandwich Place   
1             0.0   Mediterranean Restaurant           Golf Course   
2             0.0       Fast Food Restaurant           Pizza Place   
3             3.0                 Playground           Yoga Studio   
4             0.0                Coffee Shop   Sporting Goods Shop   
5             0.0                       Café           Coffee Shop   
6             0.0  Middle Eastern Restaurant        Sandwich Place   
7             0.0              Grocery Store        Sandwich Place   
8             0.0                Pizza Place   Empanada Restaurant   
9             0.0           Airport Terminal        Airport Lounge   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0      Italian Restaurant                  Café          Burger Joint   
1                 Dog Run                  Pool             Drugstore   
2    Gym / Fitness Center    Athletics & Sports             Gastropub   
3               Drugstore    Dim Sum Restaurant                 Diner   
4  Furniture / Home Store          Burger Joint            Kids Store   
5     American Restaurant                Bakery               Brewery   
6             Auto Garage        Breakfast Spot         Shopping Mall   
7            Liquor Store  Fast Food Restaurant            Beer Store   
8             Yoga Studio            Donut Shop    Dim Sum Restaurant   
9           Boat or Ferry              Boutique       Harbor / Marina   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0    Japanese Restaurant             Juice Bar        Ice Cream Shop   
1     Dim Sum Restaurant                 Diner        Discount Store   
2           Intersection              Pharmacy              Bus Line   
3         Discount Store               Dog Run      Doner Restaurant   
4     Mexican Restaurant     Fish & Chips Shop           Supermarket   
5     Italian Restaurant             Gastropub           Yoga Studio   
6  Vietnamese Restaurant                Bakery                 Diner   
7    Fried Chicken Joint              Pharmacy           Pizza Place   
8                  Diner        Discount Store               Dog Run   
9        Airport Service      Sculpture Garden          Airport Gate   

  9th Most Common Venue      10th Most Common Venue  
0       Bubble Tea Shop                 Salad Place  
1      Doner Restaurant                  Donut Shop  
2        Breakfast Spot                        Bank  
3            Donut Shop         Dumpling Restaurant  
4            Restaurant            Sushi Restaurant  
5           Fish Market              Sandwich Place  
6        Discount Stor

In [59]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    cluster = int(cluster)
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine Clusters

#### Cluster 0

In [60]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
0    Downtown Toronto             0.0                Coffee Shop   
1          North York             0.0   Mediterranean Restaurant   
2           East York             0.0       Fast Food Restaurant   
4           East York             0.0                Coffee Shop   
5        East Toronto             0.0                       Café   
6         Scarborough             0.0  Middle Eastern Restaurant   
7           Etobicoke             0.0              Grocery Store   
8          North York             0.0                Pizza Place   
9    Downtown Toronto             0.0           Airport Terminal   
11   Downtown Toronto             0.0                Coffee Shop   
12   Downtown Toronto             0.0                       Café   
13         North York             0.0             Clothing Store   
14         North York             0.0                        Pub   
16   Downtown Toronto             0.0                Coffee Shop   
17   Downtown Toronto             0.0                Coffee Shop   
19    Central Toronto             0.0             Sandwich Place   
21         North York             0.0                Coffee Shop   
22   Downtown Toronto             0.0                Coffee Shop   
23   Downtown Toronto             0.0                Coffee Shop   
24   Downtown Toronto             0.0                Coffee Shop   
25   Downtown Toronto             0.0                Coffee Shop   
26   Downtown Toronto             0.0                Coffee Shop   
27   Downtown Toronto             0.0                       Café   
29       West Toronto             0.0                        Bar   
30       West Toronto             0.0                   Pharmacy   
31          Etobicoke             0.0                       Park   
32          Etobicoke             0.0             Sandwich Place   
35          Etobicoke             0.0                  Pet Store   
37       Queen's Park             0.0                Coffee Shop   
38          Etobicoke             0.0             Discount Store   
41   Downtown Toronto             0.0              Grocery Store   
42          Etobicoke             0.0                Pizza Place   
43       East Toronto             0.0                Yoga Studio   
44        Mississauga             0.0                      Hotel   
45   Downtown Toronto             0.0                Coffee Shop   
46       West Toronto             0.0                Coffee Shop   
47       West Toronto             0.0                  Gift Shop   
48       West Toronto             0.0                       Café   
49               York             0.0              Grocery Store   
50               York             0.0                     Museum   
51         North York             0.0                       Park   
52       West Toronto             0.0                  Nightclub   
53          Etobicoke             0.0               Liquor Store   
54   Downtown Toronto             0.0                Coffee Shop   
55   Downtown Toronto             0.0                       Café   
56   Downtown Toronto             0.0                Coffee Shop   
58         North York             0.0           Sushi Restaurant   
61         North York             0.0                       Café   
62         North York             0.0             Clothing Store   
64        Scarborough             0.0       Fast Food Restaurant   
66         North York             0.0                   Pharmacy   
67        Scarborough             0.0                   Pharmacy   
68        Scarborough             0.0          Indian Restaurant   
69        Scarborough             0.0            College Stadium   
70        Scarborough             0.0        American Restaurant   
71        Scarborough             0.0                     Bakery   
72        Scarborough             0.0             Discount Store   
73        Scarborough             0.0        Fried Chicken Joint   
74

#### Cluster 1

In [61]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
28        York             1.0                  Park         Women's Store   
33        York             1.0                  Park           Yoga Studio   
57  North York             1.0                  Park                  Bank   
79  North York             1.0                  Park                  Pool   
91   East York             1.0                  Park         Metro Station   
98  North York             1.0                  Park               Airport   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
28  Fast Food Restaurant          Gourmet Shop      Department Store   
33   Dumpling Restaurant                 Diner        Discount Store   
57         Moving Target     Convenience Store           Yoga Studio   
79     Food & Drink Shop           Yoga Studio            Donut Shop   
91     Convenience Store           Yoga Studio             Drugstore   
98           Snack Place             Drugstore                 Diner   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
28  Ethiopian Restaurant   Empanada Restaurant     Electronics Store   
33               Dog Run      Doner Restaurant            Donut Shop   
57             Drugstore        Discount Store               Dog Run   
79    Dim Sum Restaurant                 Diner        Discount Store   
91        Discount Store               Dog Run      Doner Restaurant   
98        Discount Store               Dog Run      Doner Restaurant   

          9th Most Common Venue       10th Most Common Venue  
28  Eastern European Restaurant          Dumpling Restaurant  
33                    Drugstore  Eastern European Restaurant  
57             Doner Restaurant                   Donut Shop  
79                      Dog Run             Doner Restaurant  
91                   Donut Shop  Eastern European Restaurant  
98                   Donut Shop  Eastern European Restaurant

#### Cluster 2

In [62]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
60  North York             2.0             Cafeteria          Dessert Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
60    Falafel Restaurant           Event Space  Ethiopian Restaurant   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
60   Empanada Restaurant     Electronics Store  Eastern European Restaurant   

   9th Most Common Venue 10th Most Common Venue  
60   Dumpling Restaurant              Drugstore

#### Cluster 3

In [63]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
3        Scarborough             3.0            Playground   
15              York             3.0                  Park   
18   Central Toronto             3.0                  Park   
65       Scarborough             3.0                  Park   
81  Downtown Toronto             3.0                  Park   
83   Central Toronto             3.0            Playground   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
3            Yoga Studio             Drugstore    Dim Sum Restaurant   
15                 Field                 Trail          Hockey Arena   
18         Jewelry Store                 Trail      Sushi Restaurant   
65            Playground           Yoga Studio            Donut Shop   
81            Playground                 Trail      Doner Restaurant   
83                 Trail           Yoga Studio            Donut Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
3                  Diner        Discount Store               Dog Run   
15            Donut Shop                 Diner        Discount Store   
18      Doner Restaurant    Dim Sum Restaurant                 Diner   
65    Dim Sum Restaurant                 Diner        Discount Store   
81          Dessert Shop    Dim Sum Restaurant                 Diner   
83    Dim Sum Restaurant                 Diner        Discount Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
3       Doner Restaurant            Donut Shop    Dumpling Restaurant  
15               Dog Run      Doner Restaurant              Drugstore  
18        Discount Store               Dog Run             Donut Shop  
65               Dog Run      Doner Restaurant    Dumpling Restaurant  
81        Discount Store               Dog Run             Donut Shop  
83               Dog Run      Doner Restaurant              Drugstore

#### Cluster 4

In [64]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
34  North York             4.0        Baseball Field           Yoga Studio   
39   Etobicoke             4.0        Baseball Field           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
34    Dim Sum Restaurant    Falafel Restaurant           Event Space   
39    Dim Sum Restaurant    Falafel Restaurant           Event Space   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
34  Ethiopian Restaurant   Empanada Restaurant     Electronics Store   
39  Ethiopian Restaurant   Empanada Restaurant     Electronics Store   

          9th Most Common Venue 10th Most Common Venue  
34  Eastern European Restaurant    Dumpling Restaurant  
39  Eastern European Restaurant    Dumpling Restaurant

#### Cluster 5

In [65]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 5, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
10  Scarborough             5.0  Fast Food Restaurant          Dessert Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
10    Falafel Restaurant           Event Space  Ethiopian Restaurant   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
10   Empanada Restaurant     Electronics Store  Eastern European Restaurant   

   9th Most Common Venue 10th Most Common Venue  
10   Dumpling Restaurant              Drugstore

#### Cluster 6

In [66]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 6, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
40  Etobicoke             6.0                  Park                 River   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
40           Yoga Studio            Donut Shop    Dim Sum Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
40                 Diner        Discount Store               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
40      Doner Restaurant              Drugstore

#### Cluster 7

In [67]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 7, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
20  Central Toronto             7.0                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
20           Yoga Studio            Donut Shop    Dim Sum Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
20                 Diner        Discount Store               Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
20      Doner Restaurant             Drugstore       Department Store